### PR TITLE
Fix #2771, #2880

### LINKED ISSUE #2771: Make Share panels match spec/not overflow the window on shot index page
we need to write a bit of js logic to do the following:
1. ensure that share panels close onMouseOut of the parent shot
2. move the position of the panels around depending on where they are on the page so that they do not get hidden outside the window

### PR DESCRIPTION
**Pull request by niharikak101**
This is a work-in-progress for #2771 and #2880(?).

### CODE REVIEW

File: server/src/pages/shotindex/view.js [hunk 1]
```diff
@@ -52,8 +52,7 @@ class Body extends React.Component {
   renderShots() {
     let children = [];
     for (let shot of this.props.shots) {
-      let downloadUrl = this.props.downloadUrls[shot.id];
-      children.push(this.renderShot(shot, downloadUrl));
+      children.push(<Card shot = {shot} downloadUrl = {this.props.downloadUrls[shot.id]} abTests={this.props.abTests} clipUrl={shot.urlDisplay} isOwner={this.props.isOwner} staticLink={this.props.staticLink} isExtInstalled={this.props.isExtInstalled} />);
```
Comment (ianb) [inline server/src/pages/shotindex/view.js]: you shouldn't have spaces around the `=` here, the style is more like HTML

File: server/src/pages/shotindex/view.js [hunk 2]
```diff
@@ -110,19 +109,64 @@ class Body extends React.Component {
...
+      document.body.classList.remove("search-results-loading");
     }
-    return "square";
   }
 
-  renderShot(shot, downloadUrl) {
+}
+
+class Card extends React.Component {
+  constructor(props) {
```
Comment (ianb) [inline server/src/pages/shotindex/view.js]: You don't need any constructor (here you are just delegating to the superclass, but that's what happens automatically when a subclass doesn't define a method)

**Comment by niharikak101** [comment]
Update:
1. Changes the position of Share Panels if they're not in the viewport.
2. Closes the share panels onMouseLeave of the share panels, but since we are keeping them open now, shall I remove this?

**Comment by ianb** [comment]
Yes, I think remove onMouseLeave. The approach we talked about earlier should work – have the ShareButton inform its parent (via some handler passed through props) about its open/close state, then have the parent shot tile add or remove a class (e.g., `share-panel-open`) based on that state. Then change the `something:hover` selector to be `something:hover, something.share-panel-open`

**Comment by niharikak101** [comment]
Update:
1. Changes the position of Share Panels if they aren't in the viewport (improved)
2. Makes sure that the opened share panels remain open when they aren't being hovered on

File: server/src/share-buttons.js [hunk 3]
```diff
@@ -116,7 +117,21 @@ class ShareButtonPanel extends React.Component {
     </div>;
   }
 
+  changePanelPosition() {
+    let el = this.refs.share;
+    let rect = el.getBoundingClientRect();
+    if (!(rect.right <= (window.innerWidth || document.documentElement.clientWidth)) && !(rect.bottom <= (window.innerHeight || document.documentElement.clientHeight))) {
+      this.setState({left: -140, top: -385});
+    }
```
Comment (ianb) [inline server/src/share-buttons.js]: Don't the two conditionals below handle the same thing as this one check?

**Pull request merged**